OSU-Sustainability-Office/energy-dashboard
# Pacific Power Recent Data GET Endpoint

## Related Issue
- https://github.com/OSU-Sustainability-Office/automated-jobs/issues/42

## Changes
SQL code (`backend\dependencies\nodejs\models\pacific_power_data.js`)
- Unix timestamp, pacific power meter ID for each pacific power meter, in the past 3 days. If there are multiple results per meter, get the most recent one. Sort final list of results by timestamp ascending (earliest first).
- You can test the SQL query in SQL workbench with some other arbitrary timestamp (like from March 3 or something) to confirm at least 1 timestamp will be different

JS Code (`backend\dependencies\nodejs\models\pacific_power_data.js`)
- See https://playcode.io/1804875 for quick testing

## Testing
- Run backend locally (`sam local start-api` from backend directory)
- Visit http://localhost:3000/pprecent
![image](https://github.com/OSU-Sustainability-Office/energy-dashboard/assets/82061589/d2cdbcfc-7dbf-49e8-82b0-70490886968c)
- See earlier sections for other ways to test SQL and JS code

## TODO
- Update PacificPower scraper in automated-jobs repo to handle uploading missing data when detected (based on pprecent endpoint), handle case of getting previous month's data
- Cloudwatch alerts for new meters (reference ppRecent)
  - Or move this to blacklisted meters data structure
  - May cause false alarms in rare case that a meter is down for more than 3 days and then up again
- 3 days is just a placeholder value, might raise this depending on performance (needs testing)
  - Just need *some* kind of upper limit to prevent infinite scaling as more data is added (will probably get better after pacific power meter blacklist is implemented)
